django/db/backends/sqlite3/_functions.py

```diff
@@ -0,0 +1,471 @@
+    elif lookup_type == 'month':
+        return "%i-%02i-01 00:00:00" % (dt.year, dt.month)
+    elif lookup_type == 'week':
+        dt = dt - timedelta(days=dt.weekday())
+        return "%i-%02i-%02i 00:00:00" % (dt.year, dt.month, dt.day)
+    elif lookup_type == 'day':
+        return "%i-%02i-%02i 00:00:00" % (dt.year, dt.month, dt.day)
+    elif lookup_type == 'hour':
+        return "%i-%02i-%02i %02i:00:00" % (dt.year, dt.month, dt.day, dt.hour)
+    elif lookup_type == 'minute':
+        return "%i-%02i-%02i %02i:%02i:00" % (dt.year, dt.month, dt.day, dt.hour, dt.minute)
+    elif lookup_type == 'second':
+        return "%i-%02i-%02i %02i:%02i:%02i" % (dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second)
+
+
+def _sqlite_time_extract(lookup_type, dt):
+    if dt is None:
+        return None
+    try:
+        dt = typecast_time(dt)
+    except (ValueError, TypeError):
+        return None
+    return getattr(dt, lookup_type)
+
+
+def _sqlite_prepare_dtdelta_param(conn, param):
+    if conn in ['+', '-']:
+        if isinstance(param, int):
+            return timedelta(0, 0, param)
+        else:
+            return typecast_timestamp(param)
+    return param
+
+
+def _sqlite_format_dtdelta(conn, lhs, rhs):
+    """
+    LHS and RHS can be either:
+    - An integer number of microseconds
+    - A string representing a datetime
+    - A scalar value, e.g. float
+    """
+    if lhs is None or rhs is None:
+        return None
+    conn = conn.strip()
+    try:
+        real_lhs = _sqlite_prepare_dtdelta_param(conn, lhs)
+        real_rhs = _sqlite_prepare_dtdelta_param(conn, rhs)
+    except (ValueError, TypeError):
+        return None
+    if conn == '+':
+        # typecast_timestamp returns a date or a datetime without timezone.
+        # It will be formatted as "%Y-%m-%d" or "%Y-%m-%d %H:%M:%S[.%f]"
+        out = str(real_lhs + real_rhs)
+    elif conn == '-':
+        out = str(real_lhs - real_rhs)
+    elif conn == '*':
+        out = real_lhs * real_rhs
+    else:
+        out = real_lhs / real_rhs
+    return out
+
+
+def _sqlite_time_diff(lhs, rhs):
+    if lhs is None or rhs is None:
+        return None
+    left = typecast_time(lhs)
+    right = typecast_time(rhs)
+    return (
+        (left.hour * 60 * 60 * 1000000) +
+        (left.minute * 60 * 1000000) +
+        (left.second * 1000000) +
+        (left.microsecond) -
+        (right.hour * 60 * 60 * 1000000) -
+        (right.minute * 60 * 1000000) -
+        (right.second * 1000000) -
+        (right.microsecond)
+    )
+
+
+def _sqlite_timestamp_diff(lhs, rhs):
+    if lhs is None or rhs is None:
+        return None
+    left = typecast_timestamp(lhs)
+    right = typecast_timestamp(rhs)
+    return duration_microseconds(left - right)
+
+
+def _sqlite_regexp(re_pattern, re_string):
+    if re_pattern is None or re_string is None:
+        return None
+    return bool(re_search(re_pattern, str(re_string)))
```

Comment:
The difference is smaller for me, on Python 3.10 I see 90ns for `str()` and 75ns for `isinstance()`. Also you should test not just the `isinstance()` call but include the `if`:

```
In [7]: def a(text):
   ...:     return str(text)
   ...:

In [8]: def b(text):
   ...:     if not isinstance(text, str):
   ...:         text = str(text)
   ...:     return text
   ...:

In [9]: %timeit a("This is some text")
136 ns ± 1.19 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [10]: %timeit b("This is some text")
110 ns ± 0.651 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Okay, small but still faster.

> As an aside, I wonder whether we can do something to compile and cache patterns? This could make a significant difference if the function is called for a large number of rows.

There's no need as the re module uses an LRU cache of patterns - see [`re.compile()` docs](https://docs.python.org/3.10/library/re.html#re.compile):

> **Note:** The compiled versions of the most recent patterns passed to re.compile() and the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.